sig-security/security-audit-2021/RFP.md

```diff
@@ -0,0 +1,121 @@
+# Request for Proposal
+
+## Kubernetes Third-Party Security Audit
+
+The Kubernetes Third-Party Audit Working Group (working group, henceforth) is soliciting proposals from Information Security vendors for a comprehensive security audit of the Kubernetes Project.
+
+### Background
+
+In August of 2019, the Kubernetes Security Audit working group, in concert with the CNCF, Trail of Bits, and Atredis Partners, completed the first comprehensive security audit of the Kubernetes project’s [codebase](https://github.com/kubernetes/kubernetes/), working from version 1.13.
+
+These findings, below, paint a broad picture of Kubernetes security, as of version 1.13, and highlight some areas that warrant further, deeper, research.
+
+* [Kubernetes Security Review](../security-audit-2019/findings/Kubernetes%20Final%20Report.pdf)
+* [Attacking and Defending Kubernetes Installations](../security-audit-2019/findings/AtredisPartners_Attacking_Kubernetes-v1.0.pdf)
+* [Whitepaper](../security-audit-2019/findings/Kubernetes%20White%20Paper.pdf)
+* [Threat Model](../security-audit-2019/findings/Kubernetes%20Threat%20Model.pdf)
+
+### Project Goals and Scope
+
+This subsequent audit is intended to be the second in a series of recurring audits, each focusing on a specific aspect of Kubernetes while maintaining coverage of all aspects that have changed since the previous audit ([1.13](../security-audit-2019/findings/)).
+
+The scope of this audit is the most recent release (1.20) of the core [Kubernetes project](https://github.com/kubernetes/kubernetes).
+
+This audit will focus on the following components of Kubernetes:
```

Comment:
> This audit will focus on the following components of Kubernetes:

could you explain why you are excluding kubectl and kubeadm from the audit?
what deployment tools will be used for cluster creation?

Reply:
`kubectl` is "just" a client to kubeapi, so the majority of the appsec to be reviewed is on the kubeapi side of that connection - it enforces rbac, connection mgmt, input sanitization etc. Might be good to validate the connection itself from kubectl - ensure no chance of MITM etc.

`kubeadm` is interesting - it's definitely worth reviewing, but one could almost justify a separate audit - one audit for the security of k8s itself, the other is ensuring k8s is installed/upgraded/etc securely. LOC isn't small, but small compared to the rest of k8s so should probably just add it in.

Reply:
It's my opinion that we should keep kubeadm out of scope, though I'm admittedly on the fence.  A lot of what kubeadm is, is a set of default configurations and known-good states. Because k8s can be deployed in so many different ways, and through so many different distributions, I feel this limits the scope of the audit in a way that reduces value.

For example, if we pull kubeadm into scope, and it uses a default config that is not vulnerable, but the config used in kops is vulnerable, then we may miss a good bug. I'd rather focus the auditors on the components themselves, and all possible bad states they could be put into.

Reply:
I agree with John and Aaron, kubeadm might be good to include in a different audit with the other recommended Kubernetes deployment tools but imho not on this audit.

@neolit123 what is the risk you perceive of not reviewing those tools now? Can you provide some examples apart from the PITM mentioned by @jlk ?

Reply:
> @neolit123 what is the risk you perceive of not reviewing those tools now? Can you provide some examples apart from the PITM mentioned by @jlk ?

i was curious about the decision for the list of components in the audit and to include e.g. secrets-store-csi-driver vs kubectl and kubeadm, where the later are part of the official k8s release / distribution and used by a lot of users - i.e. we assume we give users secure defaults when we ship these tools.

if a separate / optional audit for those tools is preferred, that seems fine too.

i asked my second question:
> what deployment tools will be used for cluster creation?

because AFAIK the last audit used kubespray which had kubeadm implicit (as kubespray uses kubeadm).
but e.g. kops does not.

Reply:
@neolit123 It will be on us, the k8s community at large, and the external audit subproject group, to help guide the auditors to using the best deployment tools. That said, ultimately it's up to them and their level of comfort with the various offerings. The purpose of the audit is to evaluate core k8s, not its deployment technologies.

The decision on what components were selected was based on the judgment of individuals who contributed to the last audit as well as members of sig-security who volunteered their opinions. It's always going to be a value judgment, as there is far too much code to go deep everywhere.